### PR TITLE
Fix for #8253 "Unclosed client session" when initialization fails

### LINKED ISSUE #8253: "Unclosed client session" when initialization fails
### Describe the bug

This is actually a wrong use case.

Some parameters will be checked after the connection is established.

So the connection cannot be closed if there are incorrect parameters.

```python
# -*- coding: utf-8 -*-
import asyncio

from aiohttp import ClientSession


async def main():
    async with ClientSession(timeout=1):
        pass


if __name__ == '__main__':
    asyncio.run(main())
```

How about deferring connection creation ( [here](https://github.com/aio-libs/aiohttp/blob/aa014a9c1084368a01a79f09bb238486e0aa164a/aiohttp/client.py#L257) ) ?

### To Reproduce

as above

### Expected behavior

First check that the parameters are correct and then establish the connection.

### Logs/tracebacks

```python-traceback
Traceback (most recent call last):
  File "D:\proj\test.py", line 13, in <module>
    asyncio.run(main())
  File "C:\Users\user\Miniconda3\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\user\Miniconda3\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user\Miniconda3\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "D:\proj\test.py", line 8, in main
    async with ClientSession(timeout=1):
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\proj\.venv\Lib\site-packages\aiohttp\client.py", line 292, in __init__
    raise ValueError(
ValueError: timeout parameter cannot be of <class 'int'> type, please use 'timeout=ClientTimeout(...)'
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x000001D74DC83CD0>
```

### Python Version

```console
$ python --version
Python 3.11.6
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.3                                       
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp       
Author:                                              
Author-email:
License: Apache 2
Location: D:\proj\.venv\Lib\site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by:
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: D:\proj\.venv\Lib\site-packages
Requires:
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: D:\proj\.venv\Lib\site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Windows 10

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by NewGlad**
## What do these changes do?

Fix for the issue https://github.com/aio-libs/aiohttp/issues/8253

## Are there changes in behavior for the user?

Should not change behavior

## Is it a substantial burden for the maintainers to support this?
Nope!

## Related issue number
Fixes #8253

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

### CODE REVIEW

File: aiohttp/client.py [hunk 1]
```diff
@@ -325,6 +325,10 @@ def __init_subclass__(cls: Type["ClientSession"]) -> None:
         )
 
     def __del__(self, _warnings: Any = warnings) -> None:
+        if not hasattr(self, "closed"):
```
Comment (NewGlad) [inline aiohttp/client.py]: Not sure about the fix
Although it fixes the test, but looks dirty
Probably you will be able to propose a better solution?

Comment (Dreamsorcerer) [reply aiohttp/client.py]: Closed is a property, how can it not exist?

Comment (NewGlad) [reply aiohttp/client.py]: oops, my bad
should be `if not hasattr(self, "_connector"):`. Surprisingly the check for "closed" works also

Comment (Dreamsorcerer) [reply aiohttp/client.py]: What's the error it's supposed to be fixing?

Comment (Dreamsorcerer) [reply aiohttp/client.py]: Probably easier to just revert it and let the tests run, then we can decide on any fixes.

Comment (NewGlad) [reply aiohttp/client.py]: In order to don't touch `__del__`, I can set `self._connector = None` before raising the `ValueError`

Comment (Dreamsorcerer) [reply aiohttp/client.py]: That would seem sensible, given that it's already initialised a new connector before that code at the moment anyway.

File: tests/test_client_session.py [hunk 2]
```diff
@@ -841,3 +841,13 @@ async def test_build_url_returns_expected_url(
...
+
+
+async def test_instantiation_with_invalid_timeout_value(loop):
+    loop.set_debug(False)
+    logs = []
+    loop.set_exception_handler(lambda loop, ctx: logs.append(ctx))
+    with pytest.raises(ValueError, match="timeout parameter cannot be .*"):
+        ClientSession(timeout=1)
+    # should not have "Unclosed client session" warning
+    assert not logs
```
Comment (Dreamsorcerer) [inline tests/test_client_session.py]:
```suggestion
async def test_instantiation_with_invalid_timeout_value():
    with pytest.raises(ValueError, match="timeout parameter cannot be .*"):
        ClientSession(timeout=1)
```
Any warning will fail the test.

**Comment by Dreamsorcerer** [comment]
@NewGlad If you could follow the instructions above to create backports, that'd be great.

**Pull request merged**